### PR TITLE
Support cleargrad in optimizer hook

### LINKED ISSUE #2389: cleargrad in optimizer hooks
If `cleargrad` method is used in optimizer hooks, the optimizer raises `TypeError` exception.
```
class Freeze(object):
    name = 'Freeze'

    def __call__(self, opt):
        for param in opt.target.params():
            param.cleargrad()


model = chainer.links.Linear(1, 1)
optimizer = chainer.optimizers.SGD()
optimizer.setup(model)
optimizer.add_hook(Freeze())

optimizer.update()  # TypeError: unsupported operand type(s) for *: 'float' and 'NoneType'
```
If we use `param.grad[:] = 0` instead of `param.cleargrad`, it works well.

This is because `GradientMethod`  replace 'None' only before calling hooks.
https://github.com/pfnet/chainer/blob/master/chainer/optimizer.py#L400-L404

### PR DESCRIPTION
**Pull request by Hakuyume**
Fix #2389

### CODE REVIEW

File: chainer/optimizer.py [hunk 1]
```diff
@@ -369,6 +369,19 @@ class GradientMethod(Optimizer):
 
     """
 
+    def _replace_cleared_grads(self):
+        for name, param in self.target.namedparams():
+            if param.grad is None:
+                with cuda.get_device(param.data):
+                    xp = cuda.get_array_module(param.data)
+                    param.grad = xp.zeros_like(param.data)
+
```
Comment (niboshi) [inline chainer/optimizer.py]: Can you make this method public? By doing so, derived optimizers can override it to skip reallocation if not needed. (See the TODO below)
```
# TODO(unno): Some optimizers can skip this process if they does not
# affect to a parameter when its gradient is zero.
```

Then you can also remove this TODO.

As for the method name, `_replace_cleared_grads()` can be a bit ambiguous as a public method. Can you rename it to `reallocate_cleared_grads()`?

**Comment by Hakuyume** [comment]
@niboshi Thank you for your review.
I changed the name of method and removed TODO.

**Comment by niboshi** [comment]
Thank you for the fix!
Finally, could you add brief docstring to that method, mentioning that it can be overridden to skip allocation as needed?

**Comment by Hakuyume** [comment]
I added a docstring to `reallocate_cleared_grads`.

**Comment by niboshi** [comment]
LGTM. Thank you for contribution!

**Pull request merged**